### ISSUE TITLE
13% performance regression in MPS since d63db5234

### ISSUE THREAD

**Issue by danieldk** [pytorch/pytorch#78472]
### 🐛 Describe the bug

I have been testing PyTorch MPS support in spaCy over the last week or so. Performance has been great, though I have noted a somewhat large performance regression since yesterday. With bisecting I found that it occured in d63db52349ae3cffd6f762c9027e7363a6271d27. It's kind of hard to provide a minimal reproducer, but we are basically using the BERT model from Huggingface transformers with the following change:

```
diff --git a/src/transformers/models/bert/modeling_bert.py b/src/transformers/models/bert/modeling_bert.py
index 9da6258e9..1b7812dcf 100755
--- a/src/transformers/models/bert/modeling_bert.py
+++ b/src/transformers/models/bert/modeling_bert.py
@@ -324,7 +324,7 @@ class BertSelfAttention(nn.Module):
             past_key_value = (key_layer, value_layer)
 
         # Take the dot product between "query" and "key" to get the raw attention scores.
-        attention_scores = torch.matmul(query_layer, key_layer.transpose(-1, -2))
+        attention_scores = torch.matmul(query_layer, key_layer.contiguous().transpose(-1, -2).contiguous())
 
         if self.position_embedding_type == "relative_key" or self.position_embedding_type == "relative_key_query":
             seq_length = hidden_states.size()[1]
```

Without this change, the attention scores are incorrect ( I still have to look a bit deeper to understand what is causing this particular problem). But with the change above, we can reproduce CPU accuracies. I have been testing with out German transformer model, which is a finetuned BERT model.

* Before d63db52349ae3cffd6f762c9027e7363a6271d27: ~7700 words per second.
* After d63db52349ae3cffd6f762c9027e7363a6271d27: ~6700 words per second.

### Versions

```
Collecting environment information...
PyTorch version: 1.13.0a0+gitd63db52
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.4 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: version 3.23.2
Libc version: N/A

Python version: 3.10.4 (main, May 26 2022, 08:19:39) [Clang 13.1.6 (clang-1316.0.21.2.5)] (64-bit runtime)
Python platform: macOS-12.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy==0.950
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] torch==1.13.0a0+gitd63db52
[conda] Could not collect
```

**Comment by byronyi** [comment]
Inserting calls to `contiguous()` around transpose/view is known to have lower performance.

I guess it would be better to highlight wrong accuracy in your issue description; now it seems like a performance problem at the first glance, which it is not IMHO.

**Comment by danieldk** [comment]
I don’t agree. Even though that is an issue that should be addressed, it is still a performance regression. Doing a set of operations is slower than before.

**Comment by DenisVieriu97** [comment]
> I don’t agree. Even though needing `contiguous` is an issue that should be addressed, it is still a performance regression. Doing the same sequence of operations is slower than before.

@danieldk, could you please try the following build from this page https://hud.pytorch.org/pr/78496 and let me know if you are still seeing the issue? Latest nightly build also contains them (1.13.0.dev20220531).

If you are manually building PyTorch, sha 017b0ae9431ae3780a4eb9bf6d8865dfcd02cd92 contains these changes (https://github.com/pytorch/pytorch/pull/78496).

**Comment by danieldk** [comment]
I think the performance regression is largely gone (it's about 7600 words per second now). Furthermore, I only need one `contiguous` call now. I opened a separate issue about that yesterday (#78511).
